### PR TITLE
Add block log partition functionality

### PR DESCRIPTION
**Pull request by huangminghuang**
This PR backport https://github.com/EOSIO/eos/pull/9184 to support splitting block logs automatically based on block numbers. It also allows user to specify how many split block log files to retains. Once the limit is reached, the older block files can be move to another directory or deleted.

It also support recovering from the head block is not completely written to disk automatically which would eliminate most use cases for using eosio-blocklog to manually trim the end of block log file.

Some new options for chain plugin are added:

- blocks-log-stride: split the block log file when the head block number is the multiple of the stride. When the stride is reached, the current blog log and index will be renamed 'blocks-<start num>-<end num>.log/index' and a new current block log and index will be created with the most recent block. All files following this format will be used to construct an extended block log.
- max-retained-block-files: the maximum number of blocks files to retain so that the blocks in those files can be queried. When the number is reached, the oldest block file would be move to archive dir or deleted if the archive dir is empty. The retained block log files should not be manipulated by users.
- blocks-archive-dir: the location of the blocks archive directory (absolute path or relative to blocks dir). If the value is empty, blocks files beyond the retained limit will be deleted. All files in the archive directory are completely under user's control, i.e. they won't be accessed by nodeos anymore.
- blocks-retained-dir: the location of the blocks retained directory (absolute path or relative to blocks dir). If the value is empty, it is set to the value of blocks dir.
- fix-irreversible-blocks: When the existing block log is inconsistent with the index, allows fixing the log file automatically based on the index - that is, it will take the highest indexed block if it is valid; otherwise it will repair the block log and reconstruct the index.

Notice that all the newly added options cannot be specified together with `--block-log-retain-blocks` option.

### CODE REVIEW

File: libraries/chain/block_log.cpp
Comment (spoonincode) [inline libraries/chain/block_log.cpp:982]: Can `split_log()` be made safer so that it first tries to open a new file, and if that fails, continues logging to the current file?

File: libraries/chain/include/eosio/chain/log_data_base.hpp
Comment (spoonincode) [inline libraries/chain/include/eosio/chain/log_data_base.hpp:17]: Using a mapped file to supply log data makes it significantly harder to recover from IO errors. As is, should a log entry fail to read, it will SIGBUS the process I believe -- effectively SIGKILLing it. That seems rather extreme failure mode to me, particularly for something like the block log where it's easy to imagine such a failure could nominally be swallowed.

**Review by linh2931 — COMMENTED** [review]
It is a major rewrite of block_log.cpp. Please make sure test coverage is sufficient.

File: libraries/chain/block_log.cpp
Comment (linh2931) [reply libraries/chain/block_log.cpp:982]: Agreed. Had similar comments to #628.

File: libraries/chain/include/eosio/chain/log_catalog.hpp
Comment (spoonincode) [inline libraries/chain/include/eosio/chain/log_catalog.hpp:159]: It seems like, if I'm understanding this correctly, only a single partition is ever "loaded" at a time, and if a block is requested from a partition other then the one that is loaded, it closes the loaded partition and opens the needed partition.

For a node serving significant number of syncing peers that seems like it could result in a lot of thrashing. Tearing down a memory mapping and creating a new one is not cheap, and historically has involved actions like TLB flushes which need to halt all threads (I admit to not being up to date on the state of the art here, so maybe it's not as bad now as I remember).

But even forgetting the memory mapping behavior, I'm not sure how to resolve the thrashing of file descriptors: we can't keep them all open. A cache with simple LRU eviction? Or maybe just file descriptor thrashing isn't enough of a grind to worry about.

Because it appears we need to re-aquire resources, it seems very difficult to get the partitioned block log performance to be as good as the flat block log performance. A flat unpruned block log can serve multiple threads without any locking and with only a single resource acquisition at startup, for example.

File: plugins/chain_plugin/chain_plugin.cpp
Comment (spoonincode) [inline plugins/chain_plugin/chain_plugin.cpp:248]: This doesn't seem like something that should be a nodeos option, this looks like a leap-util thing

File: libraries/chain/block_log.cpp
Comment (huangminghuang) [reply libraries/chain/block_log.cpp:10]: It won't compile.

Comment (spoonincode) [reply libraries/chain/block_log.cpp:10]: `__BYTE_ORDER__` is not defined to be 1234 on little endian. The proper way to check is via something like `#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__` (consider a `static_assert` instead too)

File: plugins/chain_plugin/chain_plugin.cpp
Comment (huangminghuang) [reply plugins/chain_plugin/chain_plugin.cpp:247]: leap-util has the option to fix corrupted block log content already. This is here for the operator who doesn't care about why the log is corrupted and just want to fix it and restart the nodeos anyway. Without the flag, the operator may have to write an extra shell script to detect the situation and use leap-util to correct the problem then restart the nodeos. The option provides the functionality without the complication of shell script.

File: libraries/chain/include/eosio/chain/log_catalog.hpp
Comment (huangminghuang) [reply libraries/chain/include/eosio/chain/log_catalog.hpp:156]: I am not exactly sure what you want me to do here. Increase or the number of cached file descriptor? Or we can wait until the PR is merged and then have some performance evaluation to decide whether it needs to be addressed?

File: libraries/chain/include/eosio/chain/block_log.hpp
Comment (huangminghuang) [reply libraries/chain/include/eosio/chain/block_log.hpp]: done

File: libraries/chain/include/eosio/chain/log_data_base.hpp
Comment (huangminghuang) [reply libraries/chain/include/eosio/chain/log_data_base.hpp]: the code has been converted to regular file IO instead memory mapped file IO.

File: programs/eosio-blocklog/main.cpp
Comment (huangminghuang) [reply programs/eosio-blocklog/main.cpp]: done

File: tests/block_log.cpp
Comment (huangminghuang) [reply tests/block_log.cpp]: done

**Review by linh2931 — APPROVED** [review]

File: plugins/chain_plugin/chain_plugin.cpp
Comment (spoonincode) [reply plugins/chain_plugin/chain_plugin.cpp:247]: I'd like to get another opinion from at least one of @heifner @arhag or @766C6164 on this. 

While this isn't an option that performs an operation and then quits (_clearly_ leap-util), it's still an operation that sounds like some sort of offline recovery of a nodeos data file, which is in the ballpark of items we're been moving out of nodeos to leap-util.

It's also a problematic option from the standpoint of a user trying to decide whether or not to enable it.. why wouldn't I _always_ have this enabled? I searched through the entire diff and didn't find any error message that told the user when they should use the option or not.

It feels like if we detect an inconsistency in the block log, we should exit with an error message telling the user something is wrong and they should perform a recovery via leap-util. At least then they can decide what kind of recovery to perform. As is, it seems like it'll decide the recovery strategy based on whatever the option is -- an option the user probably didn't even know existed -- hope it's what they wanted!

File: libraries/chain/include/eosio/chain/log_catalog.hpp
Comment (spoonincode) [reply libraries/chain/include/eosio/chain/log_catalog.hpp:156]: It's not necessarily an actionable item, just something to get everyone thinking about the performance and reliability side effects of this feature.

Can you describe what happens here if we fail to open the file (because, say, we've exhausted all file descriptors)? Clearly it `return {}` but then what happens? Will it get any peers in to a weird state or such?

Comment (heifner) [reply libraries/chain/include/eosio/chain/log_catalog.hpp:156]: Not sure it is part of this effort, but would be nice to have a test that runs nodeos out of file descriptors and verifies it exits cleanly. Such a test would not necessarily run out of file descriptors right here which is why it likely should be a separate task.

File: plugins/chain_plugin/chain_plugin.cpp
Comment (heifner) [reply plugins/chain_plugin/chain_plugin.cpp:247]: > It feels like if we detect an inconsistency in the block log, we should exit with an error message telling the user something is wrong and they should perform a recovery via leap-util. At least then they can decide what kind of recovery to perform. As is, it seems like it'll decide the recovery strategy based on whatever the option is -- an option the user probably didn't even know existed -- hope it's what they wanted!

I agree. I think we should exit with an error message and let the user decide recovery strategy via `leap-util` or just restore from their own backup. In many cases the user is likely to restore from there own backup as a faster way to get running again instead of `nodeos` doing some long running recovery step.

File: libraries/chain/include/eosio/chain/log_catalog.hpp
Comment (greg7mdp) [reply libraries/chain/include/eosio/chain/log_catalog.hpp:156]: Also, maybe we should try to avoid this problem, by checking how many file descriptors we are allowed at startup, and not accepting connections if we get close to that number being in use.

File: libraries/chain/block_log.cpp
Comment (spoonincode) [inline libraries/chain/block_log.cpp:1514]: These should be `%u` -- the arguments are unsigned ints.

File: libraries/chain/include/eosio/chain/log_catalog.hpp
Comment (huangminghuang) [reply libraries/chain/include/eosio/chain/log_catalog.hpp:156]: When it returns `{}`, it means the current nodeos instance does not have the requested block, hence the net plugin should be able to response appropriately. The current main brnach can handle this behavior when a nodeos does not have the full block log. This PR does not alter it at all.

File: plugins/chain_plugin/chain_plugin.cpp
Comment (huangminghuang) [reply plugins/chain_plugin/chain_plugin.cpp:247]: option removed

File: libraries/chain/block_log.cpp
Comment (ScottBailey) [inline libraries/chain/block_log.cpp:83]: Is C++ casting preferable to C?
`ds.read( reinterpret_cast<char*>(&first_block_num), sizeof(first_block_num));`

Comment (huangminghuang) [reply libraries/chain/block_log.cpp]: we have byte order checking in line 11

File: libraries/chain/block_log.cpp
Comment (greg7mdp) [reply libraries/chain/block_log.cpp]: Hum, I'm pretty sure this is wrong advice and that `fc::exception` should be caught, as its `what()` implementation is different from the `std::exception` one.

Comment (ScottBailey) [reply libraries/chain/block_log.cpp]: > Hum, I'm pretty sure this is wrong advice and that `fc::exception` should be caught, as its `what()` implementation is different from the `std::exception` one.

Of course they are different. That's how virtual functions work, right? `fc::exception` is derived from `std::exception` and `what()` is a virtual function marked `override` in the derived class. Have we done something special bad here to make this not work?

Comment (greg7mdp) [reply libraries/chain/block_log.cpp]: My bad, you are right @ScottBailey , I missed the fact that it was a virtual function.

File: libraries/chain/block_log.cpp
Comment (ScottBailey) [reply libraries/chain/block_log.cpp]: This is corrected now.

**Review by ScottBailey — APPROVED** [review]
lgtm

**Pull request merged**
